File: iteration2/src/models/Course.java
package iteration2.src.models;

import java.util.*;

public class Course {
    private int courseCredit;
    private int courseECTS;
    private int givenSemester;
    private String courseName;
    private String courseCode;
    private Prerequisite prerequisiteInformation;
    private List<CourseSection> courseSections;
    private CourseType courseType;

    public Course() {
    }

   public Course(int courseCredit, int courseECTS, int givenSemester, String courseName, String courseCode,
            Prerequisite prerequisiteInformation, List<CourseSection> courseSections, CourseType courseType) {

      this.courseCredit = courseCredit;
        this.courseECTS = courseECTS;
        this.givenSemester = givenSemester;
        this.courseName = courseName;
        this.courseCode = courseCode;
        this.prerequisiteInformation = prerequisiteInformation;
        this.courseSections = courseSections;
        this.courseType = courseType;
    }

    public boolean checkPrerequisite(Student student) {
        if (prerequisiteInformation == null)
            return true;
        else if (prerequisiteInformation.getPrerequisiteOfCourses().isEmpty())
            return true;
        else
            return prerequisiteInformation.checkPrerequisiteCoursePassed(student, this);
    }

    public List<CourseSection> acquireAvailableSections() {
        List<CourseSection> availableCourseSections = new ArrayList<>();
        for (CourseSection courseSection : courseSections) {
            if (courseSection.checkAvailibilty()) {
                availableCourseSections.add(courseSection);
            }
        }
        return availableCourseSections;
    }

      public boolean equals(Course course) {
        return courseCode.equals(course.getCourseCode());
    }
  
    public Prerequisite getPrerequisiteInformation() {
        return prerequisiteInformation;
    }

    public int getCourseCredit() {
        return courseCredit;
    }

    public void setCourseCredit(int courseCredit) {
        this.courseCredit = courseCredit;
    }

    public int getCourseECTS() {
        return courseECTS;
    }

    public void setCourseECTS(int courseECTS) {
        this.courseECTS = courseECTS;
    }

    public String getCourseName() {
        return courseName;
    }

    public void setCourseName(String courseName) {
        this.courseName = courseName;
    }

    public String getCourseCode() {
        return courseCode;
    }

    public void setCourseCode(String courseCode) {
        this.courseCode = courseCode;
    }

    public void setPrerequisiteInformation(Prerequisite prerequisiteInformation) {
        this.prerequisiteInformation = prerequisiteInformation;
    }

    public List<CourseSection> getCourseSections() {
        return courseSections;
    }

    public CourseType getCourseType() {
        return courseType;
    }


    public int getGivenSemester() {
        return givenSemester;
    }
  

}

File: iteration3/src/models/Course.py
from iteration3.src.enums.CourseType import CourseType
from iteration3.src.models.CourseSection import CourseSection
from iteration3.src.models.Prerequisite import Prerequisite


class Course:

    def __init__(self, courseCredit, courseECTS, givenSemester, courseName, courseCode, prerequisiteInformation, courseSections, courseType):
        self._courseCredit = courseCredit
        self._courseECTS = courseECTS
        self._givenSemester = givenSemester
        self._courseName = courseName
        self._courseCode = courseCode
        self._prerequisiteInformation = prerequisiteInformation
        self._courseSections = courseSections
        self._courseType = courseType

    @classmethod
    def dictToObject(cls, courseDict):
        return cls(
            courseDict['courseCredit'],
            courseDict['courseECTS'],
            courseDict['givenSemester'],
            courseDict['courseName'],
            courseDict['courseCode'],
            Prerequisite.dictToObject(courseDict['prerequisiteInformation']) if courseDict['prerequisiteInformation'] else None,
            [CourseSection.dictToObject(courseSection) for courseSection in courseDict['courseSections']],
            CourseType.dictToEnum(courseDict['courseType'])
        )


    def checkPrerequisite(self, student):
        if self._prerequisiteInformation is None:
            return True
        elif not self._prerequisiteInformation.getPrerequisiteOfCourses():
            return True
        else:
            return self._prerequisiteInformation.checkPrerequisiteCoursePassed(student, self)
        
    
    def acquireAvailableSections(self):
        availableCourseSections = []
        for courseSection in self._courseSections:
            if courseSection.checkAvailability():
                availableCourseSections.append(courseSection)
        return availableCourseSections
    
    

    def equals(self, course):
        if self._courseCode == course.getCourseCode:
            return True
        else:
            return False
        
    def getCourseCode(self):
        return self._courseCode

    def getPrerequisiteInformation(self):
        return self._prerequisiteInformation
    
    def getCourseName(self):
        return self._courseName
    
    def getGivenSemester(self):
        return self._givenSemester
    
    def getCourseSections(self):
        return self._courseSections
    
    def getCourseCredit(self):
        return self._courseCredit

    def getCourseECTS(self):
        return self._courseECTS

    def getCourseType(self):
        return self._courseType


    